### PR TITLE
Recover from a coalesced connection that immediately goes unhealthy.

### PR DESCRIPTION
**Pull request by dave-r12**
https://github.com/square/okhttp/issues/3308

### CODE REVIEW

**Comment by dave-r12** [comment]
The test case (complex and crazy) should give the stack trace mentioned in the issue. Is there a way to simplify it?

While doing this, I hit a snag with JDK11+. If you run the test case with JDK8/9/10 it should give the stack trace but on JDK11 I was seeing this:
```
okhttp3.ConnectionCoalescingTest > coalescedConnectionDestroyedAfterAcquire FAILED
    javax.net.ssl.SSLPeerUnverifiedException: Hostname san.com not verified (no certificates)
        at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.kt:368)
        at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.kt:310)
        at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:178)
        at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:236)
        at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:109)
        at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:77)
        at okhttp3.internal.connection.Transmitter.newExchange$okhttp(Transmitter.kt:162)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:35)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:82)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:84)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:71)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.kt:184)
        at okhttp3.RealCall.execute(RealCall.kt:66)
        at okhttp3.ConnectionCoalescingTest.coalescedConnectionDestroyedAfterAcquire(ConnectionCoalescingTest.java:253)
```

I turned on the Java SSL debugging and discovered it was using TLS1.3 and "resuming a session." Here was the output:
```
javax.net.ssl|DEBUG|10|Test worker|2019-08-02 11:43:23.761 CDT|PreSharedKeyExtension.java:831|Resuming session:  (
Session(1564764200860|TLS_AES_128_GCM_SHA256)
)
```

I'm not sure I understand all the details around TLS1.3 but am wondering if this is something we need to be looking out for.

**Comment by yschimke** [comment]
@dave-r12 I think @swankjesse narrowed it down to this https://github.com/square/okhttp/issues/4703

So a bug on 11.0.2, hopefully fixed in 11.0.3?

https://bugs.openjdk.java.net/browse/JDK-8212885
https://bugs.openjdk.java.net/browse/JDK-8211806

**Comment by yschimke** [comment]
This makes sense.  It makes me a little bit worried there is an infinite loop where two concurrent requests both establish connections, find it's bad and retry the same route.  But it seems unlikely because the race winner presumably fails the call and goes back to the calling code.

**Comment by dave-r12** [comment]
@yschimke Ahh -- yes updating to the latest JDK worked. Thank you!

Ohh right. Yeah I guess this situation could go on infinitely. Hmm.

File: okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java [hunk 1]
```diff
@@ -151,6 +152,107 @@
...
+    CountDownLatch latch1 = new CountDownLatch(1);
+    CountDownLatch latch2 = new CountDownLatch(1);
+    CountDownLatch latch3 = new CountDownLatch(1);
+    CountDownLatch latch4 = new CountDownLatch(1);
+    EventListener listener1 = new EventListener() {
+      @Override public void connectStart(Call call, InetSocketAddress inetSocketAddress,
+          Proxy proxy) {
+        try {
+          // Wait for request2 to guarantee we make 2 separate connections to the server.
+          latch1.await();
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java]: rename `latch1` to `request1ConnectStart` ?

File: okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java [hunk 2]
```diff
@@ -151,6 +152,107 @@
...
+    };
+
+    EventListener request2Listener = new EventListener() {
+      @Override public void connectStart(Call call, InetSocketAddress inetSocketAddress,
+          Proxy proxy) {
+        // Let request1 proceed to make a connection.
+        latch1.countDown();
+        try {
+          // Wait until request1 makes the connection and puts it in the connection pool.
+          latch2.await();
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java]: rename `latch2` to `request1ConnectionAcquired` ?

File: okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java [hunk 3]
```diff
@@ -151,6 +152,107 @@
...
+        } catch (InterruptedException e) {
+          throw new AssertionError(e);
+        }
+      }
+
+      @Override public void connectionAcquired(Call call, Connection connection) {
+        // We obtained the coalesced connection. Let request1 violently destroy it.
+        latch3.countDown();
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java]: `latch3` -> `request2ConnectionAcquired`

File: okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java [hunk 4]
```diff
@@ -151,6 +152,107 @@
...
+      @Override public void onResponse(Call call, Response response) throws IOException {
+        try {
+          // Wait until request2 acquires the connection before we destroy it violently.
+          latch3.await();
+        } catch (InterruptedException e) {
+          throw new AssertionError(e);
+        }
+        assert200Http2Response(response, "san.com");
+        connection.get().socket().close();
+        latch4.countDown();
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/ConnectionCoalescingTest.java]: `latch4` -> `request1SocketClosed`

**Comment by dave-r12** [comment]
@swankjesse did you want me to go ahead and make those changes you suggested? You merged before I had a chance 😄.

**Comment by swankjesse** [comment]
Sure! Though it's pretty minor. I'm more interested in fuzzing race conditions.

**Pull request merged**